### PR TITLE
Added option to change collection item behavior

### PR DESCRIPTION
**Pull request by OnSive**
Added a new option `CollectionItemsShouldNotBeProxies` which is by default `true` (to not break system already using this).\
This option allows users to choose the behavior of collection item tracking.

With this option true, all items of a collection are being converted to proxies on `.AsTrackable()` like bevor.\
But if you set the option to `false`, only the collection items in the trackable object are beeing proxied. The original items in the source list will stay the same class and not be proxied - like the rest of objects with tracking.

```
ChangeTrackingFactory.Default.CollectionItemsShouldNotBeProxies = true; // Old System

var myList        = new List<MyClass>();
var myClassItem   = myList.First();        // --> typeof myClass
var trackableList = myList.AsTrackable();

var myClassItem2  = myList.First();        // --> typeof Proxy
var proxyItem     = trackableList.First(); // --> typeof Proxy
```
```
ChangeTrackingFactory.Default.CollectionItemsShouldNotBeProxies = false; // New System

var myList        = new List<MyClass>();
var myClassItem   = myList.First();        // --> typeof myClass
var trackableList = myList.AsTrackable();

myClassItem       = myList.First();        // --> typeof myClass
var proxyItem     = trackableList.First(); // --> typeof Proxy
```

### CODE REVIEW

**Comment by joelweiss** [comment]
Hi,

Ultimately both collections will are still sharing the same target instance, and changes on the target might or might not be reflected in the proxy, which is very confusing, so the intention was once proxied to not have access to the target.

 would you please explain why you need this feature?

**Comment by Int32Overflow** [comment]
@joelweiss 
Currently, lists also change the originals, which makes no sense in my opinion. The orginal object should not be change when I "track" it.
The above shows only one problem. Another one occurs when I monitor an object with child objects. 

```
    public class Data
    {
        public virtual IList<SubItem> Items { get; set; }
        public virtual string Text { get; set; }
    }
    public class Item
    {
        public virtual string Text { get; set; }
    }
```
Again, the type of items changes to a proxy in the original object.
That is, even if I do GetOrginal() on the "Data" class, the "Items" are still a proxy object. 

In a nutshell:
- I expect, when I call the method "GetOrginal", that the object with child objects comes out exactly as I threw it in.
- I expect that the original object with child objects will not get a type change. (Because that also causes that the Orginal object can't be tracked again, because the child objects are still a proxy).

**Comment by joelweiss** [comment]
Still not convinced. And it will be inconsistent with when you make trackable an object (or collection of objects) that have a `Collection` property, where the Property will always be trackable regardless of the `CollectionItemsShouldNotBeProxies` setting


@Int32Overflow 
> I expect, when I call the method "GetOrginal", that the object with child objects comes out exactly as I threw it in.

if there is an issue of `GetOriginal()` not working as expected, please file it as a separate issue.

**Comment by Int32Overflow** [comment]
@joelweiss Try this unit test:

```
using ChangeTracking;
using System.Collections.Generic;
using Xunit;

namespace Tests
{
    public class Data
    {
        public virtual Item Item { get; set; }
        public virtual IList<Item> Items { get; set; }
        public virtual string Text { get; set; }
    }

    public class Item
    {
        public virtual string Text { get; set; }
    }

    public class UnitTest1
    {
        [Fact]
        public void Test1()
        {
            var orginal = new Data()
            {
                Text = "12",
                Items = new List<Item>() { new Item() { Text = "Items_1_12" } },
                Item = new Item() { Text = "Subitem_12" }
            };

            var tracked = orginal.AsTrackable();
            tracked.Items[0].Text = "Items_1_23";
            tracked.Item.Text = "Subitem_23";

            Assert.Equal(typeof(Data), orginal.GetType());
            Assert.Equal(typeof(Item), orginal.Item.GetType());
            Assert.Equal(typeof(List<Item>), orginal.Items.GetType());
            Assert.Equal(typeof(Item), orginal.Items[0].GetType());
        }
    }
}
```

grafik

**Comment by OnSive** [comment]
@joelweiss 
> Still not convinced. And it will be inconsistent with when you make trackable an object (or collection of objects) that have a `Collection` property, where the Property will always be trackable regardless of the `CollectionItemsShouldNotBeProxies` setting

It's currently exactly as you wrote, it's inconsistent that the original items type is changed to the proxy type while everything else has the expected behavior of not changing the original type.

Only for backwards combability and not adding breaking changes, I choose to create a new option instead of just fixing the miss behavior.

> > I expect, when I call the method "GetOrginal", that the object with child objects comes out exactly as I threw it in.
> 
> if there is an issue of `GetOriginal()` not working as expected, please file it as a separate issue.

The problem is that everywhere the items are converted to proxies, not only at `GetOriginal()`, that was just an example.

**Comment by joelweiss** [comment]
The way this library is currently designed, is to not access the original objects/collections once tracked, since the original poco object instances are actually behind the proxied class instances, that means a change to the original instance will also change the proxied instance, but the tracked instance wont know about the changes made behind its back and no change notification will occur, if you do use the original object the behavior is undefined and it may or may not bite you.

For example:
```cs
void Main()
{
    Item poco = new Item { Text = "Original Test" };    
    
    Item trackable = poco.AsTrackable();
    
    //change made through the original poco
    poco.Text = "Cheater";

    //the property of the trackable is also changed, however it is not aware of it.
    Console.WriteLine($"Trackable Text: {trackable.Text}");
    Console.WriteLine($"Trackable Changed Properties Count:{trackable.CastToIChangeTrackable().ChangedProperties.Count()}");
    Console.WriteLine($"Trackable Status: {trackable.CastToIChangeTrackable().ChangeTrackingStatus}");
}

public class Item
{
    public virtual string Text { get; set; }
}
```
outputs
```
Trackable Text: Cheater
Trackable Changed Properties Count:0
Trackable Status: Unchanged
```

if you need back a poco after making trackable, you have two methods `GetOriginal()` which returns you a poco without any (unaccepted) changes and `GetCurrent()` which returns a poco with all the current values.

HTH
Joel
